File: src/validation.py
import re
from typing import List
from src.models import TranscriptSegment, ValidatedQuote


def normalize_text(text: str) -> str:
    text = text.lower()
    text = re.sub(r"[“’”\"\']", "", text)
    text = re.sub(r"[^\w\s]", " ", text)
    text = re.sub(r"\s+", " ", text).strip()
    return text


def validate_quote(
    candidate_quote: str,
    candidate_timestamp: str,
    candidate_expert: str,
    all_segments: List[TranscriptSegment],
) -> ValidatedQuote:
    # Filter segments for this expert
    expert_segments = [
        s for s in all_segments
        if s.is_expert and (
            candidate_expert.lower() in s.expert.lower()
            or s.expert.lower() in candidate_expert.lower()
            or candidate_expert.lower() in s.market.lower()
        )
    ]

    target_seg = next((s for s in expert_segments if s.timestamp == candidate_timestamp), None)
    clean_quote = normalize_text(candidate_quote)

    if target_seg:
        clean_seg = normalize_text(target_seg.text)
        if clean_quote in clean_seg:
            return ValidatedQuote(
                quote=candidate_quote,
                verbatim_text=target_seg.text,
                expert=target_seg.expert,
                market=target_seg.market,
                timestamp=target_seg.timestamp,
                is_validated=True,
                confidence_score=1.0,
                validation_status="verified",
            )

        # Word overlap check
        quote_words = [w for w in clean_quote.split() if len(w) > 2]
        matched_words = [w for w in quote_words if w in clean_seg]
        ratio = len(matched_words) / len(quote_words) if quote_words else 0

        if ratio >= 0.70:
            return ValidatedQuote(
                quote=target_seg.text,  # ground to exact verbatim
                verbatim_text=target_seg.text,
                expert=target_seg.expert,
                market=target_seg.market,
                timestamp=target_seg.timestamp,
                is_validated=True,
                confidence_score=ratio,
                validation_status="fuzzy_matched",
            )

    # Search other segments of same expert
    for seg in expert_segments:
        clean_seg = normalize_text(seg.text)
        if clean_quote in clean_seg or clean_seg in clean_quote:
            return ValidatedQuote(
                quote=seg.text,
                verbatim_text=seg.text,
                expert=seg.expert,
                market=seg.market,
                timestamp=seg.timestamp,
                is_validated=True,
                confidence_score=0.9,
                validation_status="verified",
                warning_message=f"Adjusted timestamp from {candidate_timestamp} to {seg.timestamp} based on exact transcript source."
                if seg.timestamp != candidate_timestamp else None,
            )

    # Unverified
    return ValidatedQuote(
        quote=candidate_quote,
        expert=candidate_expert,
        market=expert_segments[0].market if expert_segments else "Unknown",
        timestamp=candidate_timestamp or "Unknown",
        is_validated=False,
        confidence_score=0.0,
        validation_status="unverified",
        warning_message="Quote could not be validated against the source transcript.",
    )

File: server/validation.ts
import { TranscriptSegment } from "./data";

export interface ValidatedQuote {
  quote: string;
  verbatimText?: string;
  expert: string;
  market: string;
  timestamp: string;
  isValidated: boolean;
  confidenceScore: number;
  validationStatus: "verified" | "fuzzy_matched" | "unverified";
  warningMessage?: string;
}

function normalize(text: string): string {
  return text
    .toLowerCase()
    .replace(/[“’”"']/g, "")
    .replace(/[.,\/#!$%\^&\*;:{}=\-_`~()]/g, " ")
    .replace(/\s+/g, " ")
    .trim();
}

/**
 * Validates whether a candidate quote exists in the source transcript segments for the specified expert and timestamp.
 */
export function validateQuote(
  candidateQuote: string,
  candidateTimestamp: string,
  candidateExpert: string,
  segments: TranscriptSegment[]
): ValidatedQuote {
  // Find matching expert segments
  const expertSegments = segments.filter(
    (s) =>
      s.isExpert &&
      (s.expert.toLowerCase().includes(candidateExpert.toLowerCase()) ||
        candidateExpert.toLowerCase().includes(s.expert.toLowerCase()) ||
        s.market.toLowerCase().includes(candidateExpert.toLowerCase()))
  );

  // First look for exact timestamp match
  const timestampSegment = expertSegments.find((s) => s.timestamp === candidateTimestamp);

  const cleanQuote = normalize(candidateQuote);

  if (timestampSegment) {
    const cleanSegment = normalize(timestampSegment.text);

    // 1. Direct substring inclusion
    if (cleanSegment.includes(cleanQuote)) {
      return {
        quote: candidateQuote,
        verbatimText: timestampSegment.text,
        expert: timestampSegment.expert,
        market: timestampSegment.market,
        timestamp: timestampSegment.timestamp,
        isValidated: true,
        confidenceScore: 1.0,
        validationStatus: "verified",
      };
    }

    // 2. High-similarity token overlap or substring match
    const quoteWords = cleanQuote.split(" ").filter((w) => w.length > 2);
    const matchedWords = quoteWords.filter((w) => cleanSegment.includes(w));
    const ratio = quoteWords.length > 0 ? matchedWords.length / quoteWords.length : 0;

    if (ratio >= 0.75) {
      return {
        quote: timestampSegment.text, // auto-ground to exact source text
        verbatimText: timestampSegment.text,
        expert: timestampSegment.expert,
        market: timestampSegment.market,
        timestamp: timestampSegment.timestamp,
        isValidated: true,
        confidenceScore: ratio,
        validationStatus: "fuzzy_matched",
      };
    }
  }

  // Fallback: search across all segments of that expert
  for (const seg of expertSegments) {
    const cleanSeg = normalize(seg.text);
    if (cleanSeg.includes(cleanQuote) || cleanQuote.includes(cleanSeg)) {
      return {
        quote: seg.text,
        verbatimText: seg.text,
        expert: seg.expert,
        market: seg.market,
        timestamp: seg.timestamp,
        isValidated: true,
        confidenceScore: 0.9,
        validationStatus: "verified",
        warningMessage:
          seg.timestamp !== candidateTimestamp
            ? `Corrected timestamp from ${candidateTimestamp} to ${seg.timestamp} based on exact transcript source.`
            : undefined,
      };
    }
  }

  // Could not be validated
  return {
    quote: candidateQuote,
    expert: candidateExpert,
    market: expertSegments[0]?.market || "Unknown",
    timestamp: candidateTimestamp || "Unknown",
    isValidated: false,
    confidenceScore: 0.0,
    validationStatus: "unverified",
    warningMessage: "Quote could not be validated against the source transcript.",
  };
}
